src/dragAndSnapManager.ts

```diff
@@ -0,0 +1,170 @@
+    });
+
+    // get the mouse cursor position at startup:
+    this._currMousePos.x = getClientX(e);
+    this._currMousePos.y = getClientY(e);
+
+    this._eventManager.listen(document, moveEventName, this._moveDrag);
+    this._eventManager.listenOnce(document, moveEventName, this._onStartDrag);
+  }
+
+  private _moveDrag = (e: MouseEvent | TouchEvent) => {
+    if (this._throttleWait) {
+      return;
+    };
+    e = e || window.event;
+    // calculate the new cursor position:
+    const deltaMousePosX = this._currMousePos.x - getClientX(e);
+    const deltaMousePosY = this._currMousePos.y - getClientY(e);
+    this._currMousePos.x = getClientX(e);
+    this._currMousePos.y = getClientY(e);
+    // set the element's new position
+    if (this._draggableContainer && this._gueClientRec) {
+      let top = parseInt(this._draggableContainer.style.top, 10);
+      let bottom = parseInt(this._draggableContainer.style.bottom, 10);
+      let right = parseInt(this._draggableContainer.style.right, 10);
+      let left = parseInt(this._draggableContainer.style.left, 10);
+
+      if (Number.isInteger(right)) {
+        right = right + deltaMousePosX;
+      } else {
+        left = left - deltaMousePosX;
+      }
+      if (Number.isInteger(bottom)) {
+        bottom = bottom + deltaMousePosY;
+      } else {
+        top = top - deltaMousePosY;
+      }
+
+      this._draggableContainer.style.top = makeStyleString(top);
+      this._draggableContainer.style.right = makeStyleString(right);
+      this._draggableContainer.style.bottom = makeStyleString(bottom);
+      this._draggableContainer.style.left = makeStyleString(left);
+    }
+
+    // handle throttling to avoid performance issues on dragging
+    this._throttleWait = true;
+    setTimeout(() => {
+      this._throttleWait = false;
+    }, DRAG_THROTTLE_MS);
+  }
+
+  private _stopDrag = () => {
+    // stop moving when mouse button is released:
+    this._onStopDrag();
+    this._eventManager.unlisten(document, DragEvents.MouseMove);
+    this._eventManager.unlisten(document, DragEvents.TouchMove);
+    if (this._draggableContainer && this._gueClientRec) {
+      const boundClientRect = this._draggableContainer.getBoundingClientRect();
+      const draggableContainerCenterX = boundClientRect.x + boundClientRect.width / 2;
+      const draggableContainerCenterY = boundClientRect.y + boundClientRect.height / 2;
+      const guiClientCenterX = this._gueClientRec.x + this._gueClientRec.width / 2;
+      const guiClientCenterY = this._gueClientRec.y + this._gueClientRec.height / 2;
+      let position: Position;
+      if (draggableContainerCenterX > guiClientCenterX) {
+        if (draggableContainerCenterY > guiClientCenterY) {
+          position = Position.BottomRight;
+          this._draggableContainer.style.top = '';
+          this._draggableContainer.style.right = '0px';
+          this._draggableContainer.style.bottom = '0px';
+          this._draggableContainer.style.left = '';
```

Comment:
Extract each position as a separate method otherwise it is too long if statement

Reply:
done